File: 560. 和为 K 的子数组.py
class Solution(object):
    def subarraySum(self, nums, k):
        """
        :type nums: List[int]
        :type k: int
        :rtype: int
        """
        result = 0
        pre_sum = 0
        n = len(nums)
        pre_sum_dict = {0: 1}
        #为什么要加0:1
        # 假设 nums = [3, 4, 7], k = 3。如果没有 {0: 1}：
        # 遍历到 3：pre_sum = 3。寻找 target = 3 - 3 = 0。
        # 此时字典是空的，找不到 0，result 仍为 0。
        for num in nums:
            pre_sum += num
            if pre_sum - k in pre_sum_dict:
                #数组中有0或负数会导致pre_sum有多个相同
                result += pre_sum_dict[pre_sum - k]
            pre_sum_dict[pre_sum] = pre_sum_dict.get(pre_sum, 0) + 1
        return result


s = Solution()
print(s.subarraySum([1, 1, 1], 2))

File: 560. 和为 K 的子数组.go
package main

func subarraySum(nums []int, k int) int {
	// 滑动窗口做法只适用于数组内全为正数
	// 前缀表: 前缀和 = 出现过的次数
	// 当前位置的前缀和减掉 k 得到的前缀和出现的次数就是子串的出现次数

	res := 0
	preSumMap := make(map[int]int, 0)
	preSum := 0
	preSumMap[0] = 1 // 前缀和为0的情况出现过一次

	for _, num := range nums {
		preSum += num

		if times, ok := preSumMap[preSum-k]; ok {
			res += times
		}

		preSumMap[preSum]++

	}

	return res
}
